www_src/pages/section-2/grid.jsx

```diff
@@ -0,0 +1,382 @@
+
+      if (isEvenHeight) {
+        cameraY -= (this.slotHeight / 2);
+      }
+    }
+
+    var newDisplayState = {
+      zoom: this.state.containerWidth / (this.slotWidth * pagesWide),
+      previousZoom: this.displayState.zoom || this.state.containerWidth / (this.slotWidth * pagesWide),
+      pagesWide: pagesWide,
+      focusedPageCoords: {x: pageX, y: pageY}
+    }
+
+    if (doCentering) {
+      newDisplayState.previousCameraX = this.displayState.cameraX;
+      newDisplayState.previousCameraY = this.displayState.cameraY;
+      newDisplayState.cameraX = cameraX;
+      newDisplayState.cameraY = cameraY;
+    }
+
+    this.displayState = newDisplayState;
+  },
+  onPageClick: function (event) {
+    // TODO : TEMP - Just rotating through zoom factors
+
+    var zoomFactor = 3.25;
+
+    if (this.displayState.pagesWide) {
+      zoomFactor = this.displayState.pagesWide === 1 ? 3.25 : 1;
+    }
+
+    this.activeTile = {
+      x: event.x,
+      y: event.y
+    };
+
+    this.calculateCameraState(event.x, event.y, zoomFactor, true);
+    this.animateCamera(true);
+  },
+  showOverview: function () {
+    // 6 UP
+
+    if (this.activeTile) {
+      this.calculateCameraState(this.activeTile.x, this.activeTile.y, 6.25, true);
+    } else {
+      this.calculateCameraState(Math.floor(this.tilesPerRow / 2 ), Math.floor(this.tilesPerCol / 2), 6.25, true);
+    }
+
+    this.animateCamera(true);
+  },
+  generateGrid: function (width, height) {
+    var grid = [];
+
+    for (var h = 0; h < height; h++) {
+      grid.push([]);
+
+      for (var w = 0; w < width; w++) {
+        grid[h].push(null);
+      }
+    }
+
+    grid[Math.floor(height / 2)][Math.floor(width / 2)] = 'EMPTY';
+
+    return grid;
+  },
+  getInitialState: function () {
+    // 20 x 20 seems to be max before display glitches
+    var layout = this.generateGrid(3, 3);
+
+    return {
+      layout: layout
+    }
+  },
+  addPageClick: function (event) {
+    var self = this;
+
+    var newLayout = this.state.layout;
+
+    newLayout[event.y][event.x] = 'EMPTY';
+
+    var width = newLayout[0].length;
+    var height = newLayout.length;
+
+    // The new page tile's x and y coords may shift if the grid is expanded
+    var newPageX = event.x;
+    var newPageY = event.y;
+
+    // Detect what grid boundaries the given tile touches
+    function edgeDetect(x, y) {
+      var boundaries = [];
+
+      if (x === 0) {
+        boundaries.push('left');
+      }
+
+      if (x === newLayout[0].length - 1) {
+        boundaries.push('right');
+      }
+
+      if (y === 0) {
+        boundaries.push('top');
+      }
+
+      if (y === newLayout.length - 1) {
+        boundaries.push('bottom');
+      }
+
+      return boundaries;
+    }
+
+    // Expand the grid in the specified direction
+    function expandGrid(direction) {
+      function buildEmptyRow(length) {
+        var row = [];
+
+        for (var i = 0; i < length; i++) {
+          row.push(null);
```

Comment:
proper pet-peeve nit: Javascript has no use for null, it already has undefined, but even that is usually unnecessary since JS also has `false`